packages/babel-core/src/transformation/index.js

```diff
@@ -55,11 +55,28 @@ export function runSync(
     ast,
   );
 
-  transformFile(file, config.passes);
-
   const opts = file.opts;
-  const { outputCode, outputMap } =
-    opts.code !== false ? generateCode(config.passes, file) : {};
+  try {
+    transformFile(file, config.passes);
+  } catch (e) {
+    e.message = `${opts.filename ?? "unknown"}: ${e.message}`;
+    if (!e.code) {
+      e.code = "BABEL_TRANSFORM_ERROR";
+    }
+    throw e;
+  }
+
+  let outputCode, outputMap;
+  try {
+    ({ outputCode, outputMap } =
+      opts.code !== false ? generateCode(config.passes, file) : {});
```

Comment:
Nit: since we don't use an expression anymore, we could rewrite this as
```js
if (opts.code !== false) {
  ({ outputCode, outputMap } = generateCode(config.passes, file))
}
```

I find it a bit more readable, but it's up to you.

Reply:
Your suggestion is better. I am too lazy to refactor code styles. :P